### PR TITLE
feat(pl): support editing pl sql for ob mysql

### PR DESCRIPTION
**Pull request by zijiacj**
#### What type of PR is this?
type-feature

#### Which issue(s) this PR fixes:

Fixes #

#### Additional documentation e.g., usage docs, etc.:

```docs

```

### CODE REVIEW

File: server/odc-service/src/main/java/com/oceanbase/odc/service/session/OdcStatementCallBack.java [hunk 1]
```diff
@@ -190,39 +196,28 @@ public List<JdbcGeneralResult> doInStatement(Statement statement) throws SQLExce
...
             }
-            Future<Void> handle = null;
-            for (SqlTuple sqlTuple : this.sqls) {
-                if (handle != null) {
-                    try {
-                        handle.get();
-                    } catch (Exception e) {
-                        // eat exception
-                    }
+            Lock lock = null;
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/session/OdcStatementCallBack.java:199]: `OdcStatementCallBack` is designed for sql execution, don't add any other logic here, please

File: server/odc-service/src/main/java/com/oceanbase/odc/service/session/model/SqlAsyncExecuteReq.java [hunk 2]
```diff
@@ -54,4 +55,8 @@ public boolean ifAddROWID() {
         }
         return this.addROWID;
     }
+
+    public boolean ifEditPLSql() {
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/session/model/SqlAsyncExecuteReq.java:59]: this class is designed for sql execution, don't add any non-related logic here, please

File: server/odc-service/src/main/java/com/oceanbase/odc/service/session/ConnectConsoleService.java [hunk 3]
```diff
@@ -303,9 +312,14 @@ public SqlAsyncExecuteResp streamExecute(@NotNull String sessionId,
                 Objects.nonNull(request.getContinueExecutionOnError()) ? request.getContinueExecutionOnError()
                         : userConfigFacade.isContinueExecutionOnError();
         boolean stopOnError = !continueExecutionOnError;
-        OdcStatementCallBack statementCallBack = new OdcStatementCallBack(sqlTuples, connectionSession,
-                request.getAutoCommit(), queryLimit, stopOnError, executeContext);
-
+        OdcStatementCallBack statementCallBack = null;
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/session/ConnectConsoleService.java:315]: this method is designed for sql execution, don't add any non-related logic here, please

File: server/odc-service/src/main/java/com/oceanbase/odc/service/session/model/SqlAsyncExecuteReq.java (hunk 2 above)
Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/session/model/SqlAsyncExecuteReq.java:59]: But the implementation of ob oracle's pl editing functionality is to call the `streamExecute` method, and I think ob mysql should do the same

File: server/odc-service/src/main/java/com/oceanbase/odc/service/session/ConnectConsoleService.java (hunk 3 above)
Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/session/ConnectConsoleService.java:315]: ob oracle implement pl editing functionality in that.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/session/OdcStatementCallBack.java (hunk 1 above)
Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/session/OdcStatementCallBack.java:199]: This is the initiation of an asynchronous task executing sql, so the lock is only effective here

File: server/odc-service/src/main/java/com/oceanbase/odc/service/session/OdcStatementCallBack.java [hunk 4]
```diff
@@ -249,6 +246,48 @@ public List<JdbcGeneralResult> doInStatement(Statement statement) throws SQLExce
...
+                    handle.get();
+                } catch (Exception e) {
+                    // eat exception
+                }
+            }
+            onExecutionStart(sqlTuple);
+            try {
+                applyConnectionSettings(statement);
+            } catch (Exception e) {
+                log.warn("Init driver statistic collect failed, reason={}", e.getMessage());
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/session/OdcStatementCallBack.java:263]: collect may collection ?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/session/ConnectConsoleService.java [hunk 5]
```diff
@@ -621,4 +634,42 @@ private Integer checkQueryLimit(Integer queryLimit) {
         return queryLimit;
     }
 
+    private OdcStatementCallBack generateEditPLSqlODCStatementCallBackForOBMysql(@NotEmpty List<SqlTuple> sqlTuples,
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/session/ConnectConsoleService.java:637]: pl editor is a special sceanrio, may extrac to seprate class to handle,
make it easy for integration-test, please add integraiton test cases

File: server/integration-test/src/test/java/com/oceanbase/odc/service/session/ConnectConsoleServiceTest.java [hunk 6]
```diff
@@ -78,21 +83,300 @@
...
+        do {
+            moreResults =
+                    consoleService.getMoreResults(testConnectionSession.getId(), sqlAsyncExecuteResp.getRequestId());
+            sqlExecuteResults.addAll(moreResults.getResults());
+        } while (!moreResults.isFinished());
+        Assert.assertTrue(sqlExecuteResults.size() == 4);
+        for (SqlExecuteResult sqlExecuteResult : sqlExecuteResults) {
+            Assert.assertTrue(sqlExecuteResult.getStatus() == SqlExecuteStatus.SUCCESS);
+        }
+        syncJdbcExecutor.execute(dropTestProcedure);
```
Comment (yizhouxw) [inline server/integration-test/src/test/java/com/oceanbase/odc/service/session/ConnectConsoleServiceTest.java:126]: move env clear logic into tearDown method ,which use `@After` annotation

File: server/integration-test/src/test/java/com/oceanbase/odc/service/session/ConnectConsoleServiceTest.java [hunk 7]
```diff
@@ -78,21 +83,300 @@
...
     private ConnectConsoleService consoleService;
     @Autowired
     private DefaultDBSessionManage defaultConnectSessionManage;
 
+    @Test
+    public void editProcedureForOBMysql_normal_successResult() throws Exception {
+        ConnectionSession testConnectionSession = TestConnectionUtil.getTestConnectionSession(ConnectType.OB_MYSQL);
+        SyncJdbcExecutor syncJdbcExecutor = testConnectionSession.getSyncJdbcExecutor(
+                ConnectionSessionConstants.CONSOLE_DS_KEY);
+        String dropTestProcedure = "DROP PROCEDURE IF EXISTS ODC_TEST_PROCEDURE;";
```
Comment (yizhouxw) [inline server/integration-test/src/test/java/com/oceanbase/odc/service/session/ConnectConsoleServiceTest.java:99]: should use IsolatedNameGenerator for generate procedureName ,
for avoid pararell conflict between test cases.

File: server/integration-test/src/test/java/com/oceanbase/odc/service/session/ConnectConsoleServiceTest.java [hunk 8]
```diff
@@ -78,21 +83,300 @@
...
+    private final String sessionId = "10000";
     @MockBean
     private ConnectSessionService sessionService;
     @Autowired
     private ConnectConsoleService consoleService;
     @Autowired
     private DefaultDBSessionManage defaultConnectSessionManage;
 
+    @Test
+    public void editProcedureForOBMysql_normal_successResult() throws Exception {
```
Comment (yizhouxw) [inline server/integration-test/src/test/java/com/oceanbase/odc/service/session/ConnectConsoleServiceTest.java:95]: it is not unit testing, it is implemenation instead.
should add DBPLModifyHelperTest instead, we may always avoid write logic in test cases.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBPLModifyHelper.java [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+        sqlAsyncExecuteReq.setSql(editPLReq.getSql());
+        sqlAsyncExecuteReq.setSplit(false);
+        List<OffsetString> sqls = sqlAsyncExecuteReq.ifSplitSqls()
+                ? SqlUtils.splitWithOffset(connectionSession, sqlAsyncExecuteReq.getSql(),
+                        sessionProperties.isOracleRemoveCommentPrefix())
+                : Collections.singletonList(new OffsetString(0, sqlAsyncExecuteReq.getSql()));
+        List<SqlTuple> sqlTuples = connectConsoleService.generateSqlTuple(sqls, connectionSession, sqlAsyncExecuteReq);
+        SqlAsyncExecuteResp response = SqlAsyncExecuteResp.newSqlAsyncExecuteResp(sqlTuples);
+        Map<String, Object> context = new HashMap<>();
+        context.put(SqlCheckInterceptor.NEED_SQL_CHECK_KEY, needSqlRuleCheck);
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBPLModifyHelper.java:176]: these are all sql execution module's inner details, you may not care about these

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBPLModifyHelper.java [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+@Component
+@SkipAuthorize("permission check inside")
+public class DBPLModifyHelper {
+    public static final String ODC_TEMPORARY_PROCEDURE = "_ODC_TEMPORARY_PROCEDURE";
+    public static final String ODC_TEMPORARY_TRIGGER = "_ODC_TEMPORARY_TRIGGER";
+    public static final String ODC_TEMPORARY_FUNCTION = "_ODC_TEMPORARY_FUNCTION";
+    public static final String OB_VERSION_SUPPORT_MULTIPLE_SAME_TRIGGERS = "4.2";
+    public static final int OBTAIN_LOCK_TIME = 3;
+    @Autowired
+    @Lazy
```
Comment (MarkPotato777) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBPLModifyHelper.java:80]: What is `@Lazy` for?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBPLModifyHelper.java [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+    private ConnectConsoleService connectConsoleService;
+    @Autowired
+    private JdbcLockRegistry jdbcLockRegistry;
+    @Autowired
+    private SessionProperties sessionProperties;
+    @Autowired
+    private SqlExecuteInterceptorService sqlInterceptService;
+
+    public EditPLResp editPL(@NotNull ConnectionSession connectionSession, @NotNull @Valid EditPLReq editPLReq,
+            boolean needSqlRuleCheck)
```
Comment (MarkPotato777) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBPLModifyHelper.java:90]: Why does it need the `boolean needSqlRuleCheck` parameter?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBPLModifyHelper.java [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+    private EditPLResp executeWrappedEditPL(ConnectionSession connectionSession, EditPLReq editPLReq,
+            EditPLResp editPLResp,
+            DBObjectType plType) throws InterruptedException {
+        String plName = editPLReq.getPlName();
+        String editPLSql = editPLReq.getSql();
+        String tempPLSql = editPLSql.replaceFirst(plName, ODC_TEMPORARY_PROCEDURE);
+        SyncJdbcExecutor syncJdbcExecutor = connectionSession.getSyncJdbcExecutor(
+                ConnectionSessionConstants.CONSOLE_DS_KEY);
+        Lock editPLLock = obtainEditPLLock(connectionSession, plType);
+        if (!editPLLock.tryLock(OBTAIN_LOCK_TIME, TimeUnit.SECONDS)) {
```
Comment (MarkPotato777) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBPLModifyHelper.java:130]: Why do we need the distributed lock here?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBPLModifyHelper.java (hunk 11 above)
Comment (ysjemmm) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBPLModifyHelper.java:88]: why？（only test）

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/model/EditPLResp.java [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+ * @description:
+ * @author: zijia.cj
+ * @date: 2024/10/18 14:11
+ * @since: 4.3.3
+ */
+@Data
+public class EditPLResp {
+    // if pre-check is not successful ,sql confirmation window needs to display wrapped Sql instead of
+    // the original sql.Because what really executes is wrapped Sql
+    private String WrappedSql;
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/db/model/EditPLResp.java:36]: wrappedSql

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/model/EditPLResp.java [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+    // if shouldIntercepted is true，it indicates sql pre-check does not pass. sql confirmation window
+    // needs to be opened
+    private boolean shouldIntercepted;
+    private List<Rule> violatedRules;
+    private List<SqlTuplesWithViolation> sqls;
+    private List<UnauthorizedDBResource> unauthorizedDBResources;
+    // if none, the modification succeeded.If not none, the modification failed.it indicates the error
+    // message
+    private String failMessage;
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/db/model/EditPLResp.java:45]: errorMessage, or failedReason

File: server/odc-service/src/main/java/com/oceanbase/odc/service/session/ConnectConsoleService.java [hunk 15]
```diff
@@ -466,7 +466,7 @@ public InputStream readBinaryData(String sessionId, String sqlId, Long rowNum, I
      * Rewrite sqls, will do <br>
      * 1. add ODC_INTERNAL_ROWID query column
      */
-    private List<SqlTuple> generateSqlTuple(List<OffsetString> sqls, ConnectionSession session,
+    public List<SqlTuple> generateSqlTuple(List<OffsetString> sqls, ConnectionSession session,
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/session/ConnectConsoleService.java:469]: if change method scope to public, parameter verification should involved

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/model/EditPLResp.java [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+ * @since: 4.3.3
+ */
+@Data
+public class EditPLResp {
+    // if pre-check is not successful ,sql confirmation window needs to display wrapped Sql instead of
+    // the original sql.Because what really executes is wrapped Sql
+    private String WrappedSql;
+    // if shouldIntercepted is true，it indicates sql pre-check does not pass. sql confirmation window
+    // needs to be opened
+    private boolean shouldIntercepted;
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/db/model/EditPLResp.java:39]: we may not add UI behavior side field in model, but state side,
may violationDetected

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/model/EditPLReq.java [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+ * @author: zijia.cj
+ * @date: 2024/10/18 13:00
+ * @since: 4.3.3
+ */
+@Data
+public class EditPLReq {
+    @NotBlank
+    private String sql;
+    @NotBlank
+    private String plName;
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/db/model/EditPLReq.java:36]: objectName, objectType

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/model/EditPLResp.java (hunk 16 above)
Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/db/model/EditPLResp.java:39]: image
The original streamexecute response result also adds the status of whether to intercept. In the figure above, requestId==null indicates that an intercept is required

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/model/EditPLReq.java (hunk 17 above)
Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/db/model/EditPLReq.java:36]: ok

File: server/odc-service/src/main/java/com/oceanbase/odc/service/session/ConnectConsoleService.java (hunk 15 above)
Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/session/ConnectConsoleService.java:469]: ok

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/model/EditPLResp.java (hunk 14 above)
Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/db/model/EditPLResp.java:45]: ok

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/model/EditPLResp.java (hunk 13 above)
Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/db/model/EditPLResp.java:36]: ok

File: server/odc-service/src/main/java/com/oceanbase/odc/service/session/ConnectConsoleService.java (hunk 15 above)
Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/session/ConnectConsoleService.java:469]: Forgot to switch back to `private`

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBPLModifyHelper.java [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+ */
+@Validated
+@Component
+@SkipAuthorize("permission check inside")
+public class DBPLModifyHelper {
+    public static final String ODC_TEMPORARY_PROCEDURE = "_ODC_TEMPORARY_PROCEDURE";
+    public static final String ODC_TEMPORARY_TRIGGER = "_ODC_TEMPORARY_TRIGGER";
+    public static final String ODC_TEMPORARY_FUNCTION = "_ODC_TEMPORARY_FUNCTION";
+    public static final String OB_VERSION_SUPPORT_MULTIPLE_SAME_TRIGGERS = "4.2";
+    public static final int OBTAIN_LOCK_TIME = 3;
```
[the remark this answers is not in the document]
Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBPLModifyHelper.java:66]: ok

**Review by yizhouxw — DISMISSED** [review]
lgtm

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBPLModifyHelper.java [hunk 19]
```diff
@@ -0,0 +1,167 @@
...
+        String plName = editPLReq.getObjectName();
+        String editPLSql = editPLReq.getSql();
+        String tempPLSql = editPLSql.replaceFirst(plName, tempPlName);
+        StringBuilder wrappedSqlBuilder = new StringBuilder();
+        wrappedSqlBuilder.append("DELIMITER $$\n")
+                .append(tempPLSql).append(" $$\n")
+                .append("drop ").append(plType).append(" if exists ").append(tempPlName).append(" $$\n")
+                .append("drop ").append(plType).append(" if exists ").append(plName).append(" $$\n")
+                .append(editPLSql).append(" $$\n")
+                .append("DELIMITER ;");
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBPLModifyHelper.java:108]: why should we set the delimiter to ';'?

we should record the current delimiter, and set to it when we finish the sql execution

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBPLModifyHelper.java [hunk 20]
```diff
@@ -0,0 +1,167 @@
...
+            editPLLock.unlock();
+        }
+    }
+
+    private Lock obtainEditPLLock(@NotNull ConnectionSession connectionSession, @NonNull DBObjectType plType) {
+        ConnectionConfig connConfig =
+                (ConnectionConfig) ConnectionSessionUtil.getConnectionConfig(connectionSession);
+        Long dataSourceId = connConfig.getId();
+        String lockKeyString;
+        switch (plType) {
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBPLModifyHelper.java:155]: why should we need this switch-block?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBPLModifyHelper.java [hunk 21]
```diff
@@ -0,0 +1,167 @@
...
+                return editPLResp;
+            }
+            AsyncExecuteResultResp moreResults;
+            List<SqlExecuteResult> results = new ArrayList<>();
+            do {
+                moreResults = connectConsoleService.getMoreResults(sessionId,
+                        sqlAsyncExecuteResp.getRequestId());
+                results.addAll(moreResults.getResults());
+            } while (!moreResults.isFinished());
+            for (SqlExecuteResult result : results) {
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBPLModifyHelper.java:138]: how about add these `List<SqlExecuteResult> results` into your EditPLResp? the front end can reuse the old logic.

btw, you can ignore this comment if the front end accept EditPLResp

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBPLModifyHelper.java (hunk 19 above)
Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBPLModifyHelper.java:108]: ok

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBPLModifyHelper.java (hunk 20 above)
Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBPLModifyHelper.java:155]: remove switch

File: server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBPLModifyHelper.java (hunk 21 above)
Comment (zijiacj) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBPLModifyHelper.java:138]: ok

**Review by yhilmare — APPROVED** [review]
lgtm

**Pull request merged**